python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py

```diff
@@ -480,56 +538,111 @@ async def select_from_candidate_replicas(
-                queue_len, accepted = t.result()
-                if accepted and queue_len < lowest_queue_len:
-                    chosen_replica_id = t.replica_id
+                queue_len = t.result()
+                result.append((t.replica, queue_len))
+                self._replica_queue_len_cache.update(r.replica_id, queue_len)
+
+        assert len(result) == len(replicas)
+        return result
+
+    async def select_from_candidate_replicas(
+        self,
+        candidates: List[ReplicaWrapper],
+        backoff_index: int,
+    ) -> Optional[ReplicaWrapper]:
+        """Chooses the best replica from the list of candidates.
+
+        If none of the replicas can be scheduled, returns `None`.
+
+        The queue length for each replica is first looked up in the local cache. If not
+        present in the cache, the replica will be actively probed and the cache updated.
+
+        Among replicas that respond within the deadline and don't have full queues, the
+        one with the lowest queue length is chosen.
+        """
+        lowest_queue_len = math.inf
+        chosen_replica_id: Optional[str] = None
+
+        not_in_cache: List[ReplicaWrapper] = []
+        if self._use_replica_queue_len_cache:
+            # Populate available queue lens from the cache.
+            for r in candidates:
+                queue_len = self._replica_queue_len_cache.get(r.replica_id)
+                # Include replicas whose queues are full as not in the cache so we will
+                # actively probe them. Otherwise we may end up in "deadlock" until their
+                # cache entries expire.
+                if queue_len is None or queue_len >= r.max_concurrent_requests:
+                    not_in_cache.append(r)
+                elif queue_len < lowest_queue_len:
                     lowest_queue_len = queue_len
+                    chosen_replica_id = r.replica_id
+        else:
+            not_in_cache = candidates
 
+        # If there is a valid replica to schedule based on the information in the
+        # cache, schedule it. Else fall back to actively probing.
         if chosen_replica_id is None:
-            return None
+            for r, queue_len in await self._probe_queue_lens(
+                not_in_cache,
+                backoff_index,
+            ):
+                if queue_len is None:
+                    # None is returned if we failed to get the queue len.
+                    continue
+
+                if (
+                    queue_len < r.max_concurrent_requests
+                    and queue_len < lowest_queue_len
+                ):
+                    lowest_queue_len = queue_len
+                    chosen_replica_id = r.replica_id
+        elif len(not_in_cache) > 0:
+            # If there are replicas without a valid cache entry, probe them in the
+            # background to populate the cache.
+            self._loop.create_task(self._probe_queue_lens(not_in_cache, backoff_index))
```

Comment:
This will also continuously probe replicas that are at `max_concurrent_queries`. Is that necessary?

Reply:
yes, otherwise if all entries are at `max_concurrent_queries` we'll be in deadlock until the cache timeout is up